File: Backend/functions/src/models/ai-plan.model.ts
/**
 * AI Planning Models
 */

/**
 * 规划上下文（用户输入）
 */
export interface PlanningContext {
  destinationName: string;
  country?: string; // 可选，如果不提供则由AI生成
  budgetLevel: 'high' | 'medium' | 'low';
  startDate: string; // YYYY-MM-DD
  endDate: string; // YYYY-MM-DD
  language?: 'zh' | 'en'; // 语言，默认zh
}

/**
 * 活动
 */
export interface Activity {
  title: string;
  startTime: string; // HH:MM
  endTime: string; // HH:MM
  location: string;
  description: string;
  estimatedCost?: number;
}

/**
 * 每日行程
 */
export interface DailyItinerary {
  dayNumber: number;
  date: string; // YYYY-MM-DD
  title: string;
  activities: Activity[];
}

/**
 * 打包物品
 */
export interface PackingItem {
  name: string;
  category: string;
  quantity?: number;
  isEssential?: boolean;
}

/**
 * 待办事项
 */
export interface TodoItem {
  title: string;
  description?: string;
  priority?: 'high' | 'medium' | 'low';
}

/**
 * AI生成的目的地计划
 */
export interface AiDestinationPlan {
  tagline: string; // 一句话特色描述
  tags: string[]; // 3个标签
  detailedDescription: string; // 详细描述
  country: string; // 国家（由AI生成）
  itineraries: DailyItinerary[]; // 行程规划
  packingItems: PackingItem[]; // 打包物品
  todoChecklist: TodoItem[]; // 待办事项
}

/**
 * AI生成请求
 */
export interface GeneratePlanRequest extends PlanningContext {}

/**
 * AI生成响应
 */
export interface GeneratePlanResponse {
  success: boolean;
  data?: AiDestinationPlan;
  usage?: {
    remaining: number;
    limit: number;
  };
  error?: {
    code: string;
    message: string;
    details?: any;
  };
  upgradeUrl?: string;
}

/**
 * 使用量信息
 */
export interface UsageInfo {
  year: number;
  used: number;
  limit: number;
  remaining: number;
  subscriptionType: 'free' | 'vip';
  resetDate: string;
}

File: App/lib/features/ai_planning/data/models/ai_plan_model.dart
/// AI Plan Models
///
/// AI生成的旅行计划相关模型

class AIPlanModel {
  final String tagline;
  final List<String> tags;
  final String description;
  final List<DailyItineraryModel> dailyItineraries;
  final List<PackingItemModel> packingItems;
  final List<TodoItemModel> todoItems;

  AIPlanModel({
    required this.tagline,
    required this.tags,
    required this.description,
    required this.dailyItineraries,
    required this.packingItems,
    required this.todoItems,
  });

  factory AIPlanModel.fromJson(Map<String, dynamic> json) {
    return AIPlanModel(
      tagline: json['tagline'] as String,
      tags: (json['tags'] as List<dynamic>).map((e) => e as String).toList(),
      description: json['description'] as String,
      dailyItineraries: (json['dailyItineraries'] as List<dynamic>)
          .map((e) => DailyItineraryModel.fromJson(e as Map<String, dynamic>))
          .toList(),
      packingItems: (json['packingItems'] as List<dynamic>)
          .map((e) => PackingItemModel.fromJson(e as Map<String, dynamic>))
          .toList(),
      todoItems: (json['todoItems'] as List<dynamic>)
          .map((e) => TodoItemModel.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

class DailyItineraryModel {
  final int dayNumber;
  final String activities;
  final String meals;
  final String notes;

  DailyItineraryModel({
    required this.dayNumber,
    required this.activities,
    required this.meals,
    required this.notes,
  });

  factory DailyItineraryModel.fromJson(Map<String, dynamic> json) {
    return DailyItineraryModel(
      dayNumber: json['dayNumber'] as int,
      activities: json['activities'] as String,
      meals: json['meals'] as String? ?? '',
      notes: json['notes'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'dayNumber': dayNumber,
      'activities': activities,
      'meals': meals,
      'notes': notes,
    };
  }
}

class PackingItemModel {
  final String name;
  final String category;
  final int quantity;

  PackingItemModel({
    required this.name,
    required this.category,
    required this.quantity,
  });

  factory PackingItemModel.fromJson(Map<String, dynamic> json) {
    return PackingItemModel(
      name: json['name'] as String,
      category: json['category'] as String,
      quantity: json['quantity'] as int? ?? 1,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'category': category,
      'quantity': quantity,
    };
  }
}

class TodoItemModel {
  final String title;
  final String? deadline;

  TodoItemModel({
    required this.title,
    this.deadline,
  });

  factory TodoItemModel.fromJson(Map<String, dynamic> json) {
    return TodoItemModel(
      title: json['title'] as String,
      deadline: json['deadline'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'title': title,
      if (deadline != null) 'deadline': deadline,
    };
  }
}

/// AI生成请求
class AIPlanRequest {
  final String destinationName;
  final String budgetLevel;
  final String startDate;
  final String endDate;

  AIPlanRequest({
    required this.destinationName,
    required this.budgetLevel,
    required this.startDate,
    required this.endDate,
  });

  Map<String, dynamic> toJson() {
    return {
      'destinationName': destinationName,
      'budgetLevel': budgetLevel,
      'startDate': startDate,
      'endDate': endDate,
    };
  }
}

/// AI使用情况
class AIUsageModel {
  final int usedCount;
  final int totalQuota;
  final int remainingQuota;
  final String subscriptionType;

  AIUsageModel({
    required this.usedCount,
    required this.totalQuota,
    required this.remainingQuota,
    required this.subscriptionType,
  });

  factory AIUsageModel.fromJson(Map<String, dynamic> json) {
    return AIUsageModel(
      usedCount: json['usedCount'] as int? ?? json['used_count'] as int,
      totalQuota: json['totalQuota'] as int? ?? json['total_quota'] as int,
      remainingQuota: json['remainingQuota'] as int? ?? json['remaining_quota'] as int,
      subscriptionType: json['subscriptionType'] as String? ?? json['subscription_type'] as String,
    );
  }

  bool get hasQuota => remainingQuota > 0;
  bool get isVip => subscriptionType == 'vip';
}
